### PR TITLE
[DOC] Updates to external listener config procedures

### PR DESCRIPTION
**Pull request by PaulRMellor**
**Documentation**

Updates to the procedures describing how to connect to the Kafka cluster using an external listener.

- Node port procedure updated to use single step retrieval of addresses used in connection
- More context on what's created when each listener config is applied
- Made the procedures more consistent

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc [hunk 1]
```diff
@@ -45,22 +52,33 @@ spec:
   zookeeper:
     # ...
 ----
+<1> Ingress hosts for the bootstrap service and kafka brokers.
 
-. Make sure the hosts in the `configuration` section properly resolve to the Ingress endpoints.
+. Make sure the hosts in the `configuration` resolve to the Ingress endpoints.
++
+[source,shell,subs=+quotes]
```
Comment (PaulRMellor) [inline documentation/modules/proc-accessing-kafka-using-ingress.adoc:59]: @scholzj - Not sure what this step is meant to do. So added some placeholder commands!

**Comment by PaulRMellor** [comment]
The procedures don't describe additional config such as overrides and DNSannotations.
Not sure we want to include _everything_

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc [hunk 2]
```diff
@@ -45,22 +52,33 @@ spec:
...
+kubectl get services
+kubectl describe service _INGRESS-NAME_
```
Comment (PaulRMellor) [inline documentation/modules/proc-accessing-kafka-using-ingress.adoc:61]: @scholzj - Not sure what this step is meant to do. So added some placeholder commands!

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc [hunk 3]
```diff
@@ -5,16 +5,23 @@
 [id='proc-accessing-kafka-using-loadbalancers-{context}']
 = Accessing Kafka using loadbalancers
 
+This procedure describes how to access a Strimzi Kafka cluster from an external client using loadbalancers.
+
+To connect to a broker, you need a hostname (advertised address) for the loadbalancer _bootstrap address_,
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:10]:
```suggestion
To connect to a broker, you need the address of the _bootstrap loadbalancer_,
```

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc [hunk 4]
```diff
@@ -5,16 +5,23 @@
...
+as well as the certificate used for authentication.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:11]:
```suggestion
as well as the certificate used for TLS encryption.
```

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc [hunk 5]
```diff
@@ -36,27 +43,28 @@ spec:
 
 . Create or update the resource.
 +
-This can be done using `kubectl apply`:
 [source,shell,subs=+quotes]
 kubectl apply -f _KAFKA-CONFIG-FILE_
++
+`loadbalancer` type services and loadbalancers are created for each Kafka broker, as well as an external _bootstrap service_.
+The bootstrap service routes external traffic to the Kafka brokers.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:50]:
```suggestion
The bootstrap service routes external traffic to all Kafka brokers.
```

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc [hunk 6]
```diff
@@ -36,27 +43,28 @@ spec:
...
+DNS names and IP addresses used for connection are propagated to the `status` of each service.
 
-. Find the hostname of the bootstrap loadbalancer.
+. Retrieve the address of the bootstrap service to access the Kafka cluster.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:53]: This whole section / bullet point should be replaced with something like this

```suggestion
. Retrieve the address of the bootstrap service to access the Kafka cluster.

[source,shell,subs=+quotes]
kubectl get kafka _KAFKA-CLUSTER-NAME_ -o=jsonpath='{.status.listeners[?(@.type=="external")].bootstrapServers}{"\n"}'
```

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 This procedure describes how to access a Strimzi Kafka cluster from an external client using node ports.
 
-To connect to a broker, you need the hostname (advertised address) and port number for the Kafka _bootstrap_ address,
+To connect to a broker, you need a hostname (advertised address) and port number for the Kafka _bootstrap address_,
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-nodeports.adoc:10]:
```suggestion
To connect to a broker, you need a hostname and port number for the Kafka _bootstrap address_,
```

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc [hunk 8]
```diff
@@ -35,36 +35,38 @@ spec:
         authentication:
           type: tls
         configuration:
-          brokerCertChainAndKey: <1>
+          brokerCertChainAndKey: <2>
             secretName: my-secret
             certificate: my-certificate.crt
             key: my-key.key
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-nodeports.adoc:41]: This is unrelated to the nodeports. MAybe we should remove it from this procedure?

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc [hunk 9]
```diff
@@ -35,36 +35,38 @@ spec:
...
+<3> Optional configuration to xref:con-kafka-broker-external-listeners-nodeports-{context}[specify a preference for the first address type used by Strimzi as the node address].
 
 . Create or update the resource.
 +
 [source,shell,subs=+quotes]
 kubectl apply -f _KAFKA-CONFIG-FILE_
++
+`NodePort` type services are created for each Kafka broker, as well as an external _bootstrap service_.
+The bootstrap service routes external traffic to the Kafka brokers.
+Node addresses used for connection are propagated to the `status` of each service.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-nodeports.adoc:57]:
```suggestion
Node addresses used for connection are propagated to the `status` of the Kafka custom resource.
```

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc [hunk 10]
```diff
@@ -35,36 +35,38 @@ spec:
...
+. Retrieve the connection details to access the Kafka cluster.
 +
 [source,shell,subs=+quotes]
-kubectl get service _KAFKA-CLUSTER-NAME_-kafka-external-bootstrap -o=jsonpath='{.spec.ports[0].nodePort}{"\n"}'
+kubectl get kafka _KAFKA-CLUSTER-NAME_ -o=jsonpath='{.status.listeners[?(@.type=="external")].bootstrapServers}{"\n"}'
 +
-The port is used in the Kafka _bootstrap_ address.
-
-. Find the address of the Kubernetes node.
+You need:
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-nodeports.adoc:64]: From this point on the rest of the `Retrieve the connection details to access the Kafka cluster.` section should be removed.

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 11]
```diff
@@ -5,16 +5,23 @@
 [id='proc-accessing-kafka-using-routes-{context}']
 = Accessing Kafka using OpenShift routes
 
+This procedure describes how to access a Strimzi Kafka cluster from an external client outside of OpenShift using routes.
+
+To connect to a broker, you need a hostname (advertised address) for the route _bootstrap address_,
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:10]:
```suggestion
To connect to a broker, you need a hostname for the route _bootstrap address_,
```

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 12]
```diff
@@ -5,16 +5,23 @@
...
+as well as the certificate used for authentication.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:11]:
```suggestion
as well as the certificate used for TLS encryption.
```

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 13]
```diff
@@ -5,16 +5,23 @@
...
+
+Routes are used for HTTP access, so the port used to connect is always 443.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:13]: Again, we should either remove the HTTP reference or explain the details about TLS passthrough masking the Kafka traffic as HTTPS.

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 14]
```diff
@@ -36,13 +43,20 @@ spec:
 +
 [source,shell,subs=+quotes]
 oc apply -f _KAFKA-CONFIG-FILE_
++
+`ClusterIP` type services are created for each Kafka broker, as well as an external _bootstrap service_.
+The bootstrap service routes external traffic to the Kafka brokers.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:48]:
```suggestion
The services route the traffic from the OpenShift Routes to the Kafka brokers.
```

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 15]
```diff
@@ -36,13 +43,20 @@ spec:
...
+An OpenShift `Route` resource is also created for each service to expose them using the HAProxy load balancer.
+DNS addresses used for connection are propagated to the `status` of each service.
 
-. Find the address of the bootstrap `Route`.
+. Retrieve the address of the bootstrap service to access the Kafka cluster.
 +
 [source,shell,subs=+quotes]
 oc get routes _KAFKA-CLUSTER-NAME_-kafka-bootstrap -o=jsonpath='{.status.ingress[0].host}{"\n"}'
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:55]: Again, you can just use the command `kubectl get kafka my-cluster -o=jsonpath='{.status.listeners[?(@.type=="external")].bootstrapServers}{"\n"}'`

File: documentation/modules/proc-accessing-kafka-using-routes.adoc [hunk 16]
```diff
@@ -36,13 +43,20 @@ spec:
...
 +
-Use the address together with port 443 in your Kafka client as the _bootstrap_ address.
+You need the address of the `Route` bootstrap service (`_KAFKA-CLUSTER-NAME_-kafka-bootstrap`).
++
+Use the route service address and port 443 (_ROUTE-ADDRESS:443_) in your Kafka client as the _bootstrap address_ to connect to the Kafka cluster.
```
Comment (scholzj) [inline documentation/modules/proc-accessing-kafka-using-routes.adoc:59]:
```suggestion
```

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc (hunk 2 above)
Comment (scholzj) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:61]: This whole segment should be removed - I do not think there is any easy way to check the routing of the DNS names to the Ingress.

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc [hunk 17]
```diff
@@ -5,19 +5,26 @@
 [id='proc-accessing-kafka-using-ingress-{context}']
 = Accessing Kafka using ingress
 
-This procedure shows how to access Strimzi Kafka clusters from outside of Kubernetes using Ingress.
+This procedure shows how to access a Strimzi Kafka cluster from an external client outside of Kubernetes using Ingress.
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:8]: Out of curiosity, isn't it right to say just "Kubernetes Ingress" and the fact that you are using the Nginx Ingress controller is only an implementation detail but maybe you could use another controller?

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc [hunk 18]
```diff
@@ -5,19 +5,26 @@
...
+
+To connect to a broker, you need a hostname (advertised address) for the Ingress _bootstrap address_,
+as well as the certificate used for authentication.
+
+Ingress is used for HTTP access, so the port used to connect is always 443.
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:13]: And should be it HTTPS because of 443 (or it's a non-encrypted 443?)

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc (hunk 17 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:8]: The Ingress approach IMHO didn't went well for Kubernetes. The controllers are not fully compatible - they work fine for simple HTTP use-cases but not for some edge stuff such as TLS passthrough etc. In Strimzi, we quite specifically say that we support only Nginx Ingress which is the only one we work with and test with right now. This is from my understanding also why the Ingress class exists etc.

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc (hunk 2 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:61]: Removed the step

File: documentation/modules/proc-accessing-kafka-using-ingress.adoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-ingress.adoc:13]: Updated. Updated the other procedures to be the same.

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc (hunk 6 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:53]: @scholzj I originally changed this and then changed back, thinking the _status_ was the key. I've replaced now, but we don't want both?

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 8 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:41]: We can remove it, but isn't it used with nodeport? https://strimzi.io/docs/operators/master/using.html#type-NodePortListenerConfiguration-reference

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 10 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:64]: Can we confirm/verify what the reader needs to pick up from the fetch? 

> Use the address and port number of one of the kafka broker nodes (_NODE-ADDRESS:NODE-PORT_) in your Kafka client as the _bootstrap address_ to connect to the Kafka cluster.

File: documentation/modules/proc-accessing-kafka-using-routes.adoc (hunk 13 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-routes.adoc:13]: > For access using routes, the port is always 443.

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc (hunk 6 above)
Comment (scholzj) [reply documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:53]: What we now basically do is we do the original steps in the operator and put it always into a single place. So instead of taking it from the status of the operator, the user can take it from the Kafka status with all the methods. So the user does not have to search for the services etc.

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 8 above)
Comment (scholzj) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:41]: It can be used with any external listener. So it is not wrong per se. But it is not required, just optional. So I think having it in this example just here might be confusing.

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 10 above)
Comment (scholzj) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:64]: When the user runs this command: `kubectl get kafka _KAFKA-CLUSTER-NAME_ -o=jsonpath='{.status.listeners[?(@.type=="external")].bootstrapServers}{"\n"}'` ... it will give back the bootstrap address ... so the `_NODE-ADDRESS:NODE-PORT_`. That is why the user doesn't need anything more.

File: documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc (hunk 6 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-loadbalancers.adoc:53]: Thanks. I get it now.

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 8 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:41]: Removed. We talk about it in the concepts and _Providing your own Kafka listener certificates_. I guess that's enough.

File: documentation/modules/proc-accessing-kafka-using-nodeports.adoc (hunk 10 above)
Comment (PaulRMellor) [reply documentation/modules/proc-accessing-kafka-using-nodeports.adoc:64]: Removed

**Comment by scholzj** [comment]
@PaulRMellor There seem to be a lot of comments which are resolved but do not seem to be really. Mostly it is the comments with the suggestions from my first review. Did it somewhere lost the commit with them?

**Comment by PaulRMellor** [comment]
> @PaulRMellor There seem to be a lot of comments which are resolved but do not seem to be really. Mostly it is the comments with the suggestions from my first review. Did it somewhere lost the commit with them?

Thanks fully, I'd stashed them. Should be up-to-date now.

**Comment by scholzj** [comment]
@PaulRMellor Could you fix the DCO as well please?

**Comment by scholzj** [comment]
@ppatierno Could you please have a look at this? If it looks good to you maybe we could include it in 0.19.0 docs as well.
